api/src/test/java/org/openmrs/api/impl/PatientServiceImplTest.java

```diff
@@ -271,28 +327,53 @@ public void getPatientIdentifierTypes_shouldReturnEmptyListGivenDaoReturnsNull()
 		assertEquals(0, actualIdentifierTypes.size());
 	}
 
+	@Test
+	public void exitFromCare_shouldThrowApiExceptionIfPatientIsNull() {
+
+
+		assertThrows(APIException.class,
+			() -> patientService.exitFromCare(null, newDate, concept));
+		verify(messageSourceServiceMock).getMessage(eq("Patient.invalid.care"), eq(null), any());
+	}
```

Comment:
Please take this suggestion into your commit as is and then adjust the other 2 test cases according to my comments below.

```suggestion
	@Test
	public void exitFromCare_shouldFailIfGivenPatientIsNull() {

		when((messageSourceService).getMessage(eq("Patient.invalid.care"), any(), any()))
			.thenReturn("localized error message");

		Exception thrown = assertThrows(APIException.class,
			() -> patientService.exitFromCare(null, new Date(), new Concept()));
		assertThat(thrown.getMessage(), is("localized error message"));
	}
```

* Name: We now what exception it throws from the assertion in the test so we can shorten the name to just contain `fail`. Please take my name as is and adjust the the test names of the other 2 cases appropriately.
* Formatting: Only add one newline right after the test name, then write your test. No newline at the end of the test method. You can in general use the style called ArrangeActAssert (AAA). First arrange any data/objects you need, insert newline, then act which means call what you want to test, insert a newline, then assert. This gives you 3 blocks that are easy to spot and read. In our test cases we can have 2 blocks the setup with the `when` then the assertions. Since assertThrows contains the Act part we only have 2 blocks.
* I am more interested in the fact that we hit the right case in the code, thats the only reason I want to assert on the error message. I would not do that otherwise. Here its important since we have 3 cases throwing the APIException so if we do not do that it could be that we have multiple tests testing the same case. I prefer using `when` here to setup what the `messageSourceService` translates the key to and then assert it actually lands in the APIException. This makes sure that the user of the API will also get this message. Asserting that `verify(messageSourceServiceMock).getMessage` only tells us that it has or has not been called but does not provide us what I just explained.

Reply:
i agree!! i was trying this first along with when+verify, i dont know why it wasnt working.... thats why i skipped the when and pushed it.... Done with this change... thanks a lot for all the explanation and guidance :)

do check and let me know what should i do next..